stan/math/rev/fun/log_sum_exp.hpp

```diff
@@ -69,70 +57,16 @@ inline var log_sum_exp(const var& a, double b) {
  * Returns the log sum of exponentials.
  */
 inline var log_sum_exp(double a, const var& b) {
-  return var(new internal::log_sum_exp_dv_vari(a, b.vi_));
-}
-
-namespace internal {
-inline double log_sum_exp_as_double(const std::vector<var>& x) {
-  using std::exp;
-  using std::log;
-  double max = NEGATIVE_INFTY;
-  for (size_t i = 0; i < x.size(); ++i) {
-    if (x[i] > max) {
-      max = x[i].val();
-    }
-  }
-  double sum = 0.0;
-  for (size_t i = 0; i < x.size(); ++i) {
-    if (x[i] != NEGATIVE_INFTY) {
-      sum += exp(x[i].val() - max);
-    }
-  }
-  return max + log(sum);
-}
-
-class log_sum_exp_vector_vari : public op_vector_vari {
- public:
-  explicit log_sum_exp_vector_vari(const std::vector<var>& x)
-      : op_vector_vari(log_sum_exp_as_double(x), x) {}
-  void chain() {
-    for (size_t i = 0; i < size_; ++i) {
-      vis_[i]->adj_ += adj_ * calculate_chain(vis_[i]->val_, val_);
-    }
-  }
-};
-}  // namespace internal
-
-/**
- * Returns the log sum of exponentials.
- */
-inline var log_sum_exp(const std::vector<var>& x) {
-  return var(new internal::log_sum_exp_vector_vari(x));
+  return var(new internal::log_sum_exp_vd_vari(b.vi_, a));
 }
 
 namespace internal {
 
-// these function and the following class just translate
-// log_sum_exp for std::vector for Eigen::Matrix
-
-template <int R, int C>
-inline double log_sum_exp_as_double(const Eigen::Matrix<var, R, C>& x) {
-  if (x.size() == 0) {
-    return NEGATIVE_INFTY;
-  }
-
-  const double max = x.val().maxCoeff();
-  if (!std::isfinite(max)) {
-    return max;
-  }
-  return max + std::log((x.val().array() - max).exp().sum());
-}
-
 class log_sum_exp_matrix_vari : public op_matrix_vari {
  public:
-  template <int R, int C>
-  explicit log_sum_exp_matrix_vari(const Eigen::Matrix<var, R, C>& x)
-      : op_matrix_vari(log_sum_exp_as_double(x), x) {}
+  template <typename T>
+  explicit log_sum_exp_matrix_vari(T&& x)
+      : op_matrix_vari(log_sum_exp(x.val()), std::forward<T>(x)) {}
```

Comment:
Perfect forwarding to a function that accepts a const& is pointless. Const ref should be used here as well.